File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	bytesSent := 0
	for bytesSent < len(bytes) {
		n, err := socket.Write(bytes[bytesSent:])
		if n > 0 {
			bytesSent += n
		}
		if err != nil {
			return err
		}
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	bytesReceived := 0

	for bytesReceived < size {
		n, err := socket.Read(buff[bytesReceived:])
		if n > 0 {
			bytesReceived += n
		}
		if bytesReceived == size {
			return buff, nil
		}
		if err != nil {
			if err == io.EOF {
				return nil, io.ErrUnexpectedEOF
			}
			return nil, err
		}
	}
	return buff, nil
}

File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation

def recv_all(socket: socket.socket, size):
    buffer = []
    bytesRecieved = 0
    while bytesRecieved < size:
        buff = socket.recv(size - bytesRecieved)   
        if not buff:
            if bytesRecieved == 0:
                return b""
            raise ConnectionError("Socket connection failed")
        buffer.append(buff)
        bytesRecieved += len(buff)
    return b''.join(buffer)

def send_all(socket: socket.socket, bytes):
    bytesSent = 0

    while bytesSent < len(bytes):
        n = socket.send(bytes[bytesSent:])

        if n is None:
            raise ConnectionError("Socket connection failed")

        bytesSent += n

    return bytesSent

